### PR TITLE
Download chart

### LINKED ISSUE #747: Make it possible to download the chart on the Data page as a png
We'd like to add a little "Download" button at the top of the chart on the Data page: [Figma](https://www.figma.com/file/0UZOuDCrwGqWLKOhAyvGbc/Forrest's-First-Figma-File-(FFFF)?type=design&node-id=1613-43920&mode=design&t=pHEGA5uIguLl9CeW-0)

When you click this button, we should convert the Recharts graph into a png, and then download it. Here's a [Stack Overflow post](https://stackoverflow.com/questions/45086005/recharts-component-to-png) about how to do this, and here's a [library](https://github.com/brammitch/recharts-to-png) we might be able to use.

We should name the graph based on the metric and breakdown, if there is one selected. So e.g. "expenses.png" or "expenses_by_type.png".

### PR DESCRIPTION
**Pull request by ivan-kishko**
## Description of the change

> Make charts downloadable

## Type of change
Type: Feature

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #747 

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Comment by ivan-kishko** [comment]
Hello there! @mxosman question to you. For this task I had to add recharts-to-png and file-saver packages. BUT if I make <code>yarn add "package name"</code>  I get this
Screenshot 2023-07-07 at 4 27 39 PM
and this
Screenshot 2023-07-07 at 4 27 06 PM
If i add package as dev dependency
 <code>yarn add "package name" --dev</code> I get only what is on a second screenshot. Not sure what is going on here maybe you know?

**Comment by mxosman** [comment]
> Hello there! @mxosman question to you. For this task I had to add recharts-to-png and file-saver packages. BUT if I make `yarn add "package name"` I get this [img] and this [img] If i add package as dev dependency `yarn add "package name" --dev` I get only what is on a second screenshot. Not sure what is going on here maybe you know?

That's strange, Ivan! Try closing out and reopening your VSCode - maybe that's in a weird state. Sometimes this has happened to me before and I've been able to get around by restarting my VS code. I'm in your branch and I did `yarn install` and get no TS errors (I removed the ESlint disable comments with no problems too).

Screenshot 2023-07-07 at 11 42 39 AM

**Comment by mxosman** [comment]
I'm still going through the code - and playing around with the png - super cool library and functionality! 

Was the goal to get just this basic raw chart (below) or something more presentable with headers (like the metric title, etc.)?
and_pretrial_funding

**Comment by lilidworkin** [comment]
> Was the goal to get just this basic raw chart (below) or something more presentable with headers (like the metric title, etc.)?

Great Q -- I think just the basic raw chart is fine for now -- but we should revisit later!

File: publisher/src/components/DataViz/MetricsDataChart.tsx [hunk 1]
```diff
@@ -63,6 +69,24 @@ export const MetricsDataChart: React.FC = observer(() => {
...
+
+    const fileName = `${
+      metricSearchParam
+        ? metricSearchParam.toLowerCase().split("_").slice(1).join("_")
+        : "metric"
+    }${
+      dataVizStore.disaggregationName.toLowerCase() === "none"
+        ? ""
+        : `_by_${dataVizStore.disaggregationName.toLowerCase()}`
+    }.png`.replaceAll(" ", "_");
```
Comment (mxosman) [inline publisher/src/components/DataViz/MetricsDataChart.tsx:83]: Can we make a helper function for this and add to the utils? Also a comment on what this string transformation will do would be helpful!

This logic might not work as intended for systems like `courts_and_pretrial`

File: publisher/src/components/DataViz/MetricsDataChart.tsx [hunk 2]
```diff
@@ -28,9 +29,13 @@ import { showToast } from "@justice-counts/common/components/Toast";
 import { useWindowWidth } from "@justice-counts/common/hooks";
 import { AgencySystems, ReportFrequency } from "@justice-counts/common/types";
 import { frequencyString } from "@justice-counts/common/utils/helperUtils";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import FileSaver from "file-saver";
 import { observer } from "mobx-react-lite";
-import React, { useEffect, useState } from "react";
+import React, { useCallback, useEffect, useState } from "react";
 import { createSearchParams, useNavigate, useParams } from "react-router-dom";
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (mxosman) [inline publisher/src/components/DataViz/MetricsDataChart.tsx:37]: So strange - I'm able to remove these with no lint errors or warnings.

File: publisher/src/components/DataViz/MetricsDataChart.tsx [hunk 3]
```diff
@@ -347,12 +371,17 @@ export const MetricsDataChart: React.FC = observer(() => {
...
+              <DownloadChartIcon />
+              Download
+            </Styled.PanelRightTopButton>
           </Styled.PanelRightTopButtonsContainer>
           <ConnectedDatapointsView
             metric={metricSearchParam}
             metricName={metricName}
             metricFrequency={metricFrequency}
             dataView={dataView}
+            ref={ref}
```
Comment (mxosman) [inline publisher/src/components/DataViz/MetricsDataChart.tsx:384]: Is this the original `ref` we're drilling down to other components (coming from `useCurrentPng`)?

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 1 above)
Comment (mxosman) [reply publisher/src/components/DataViz/MetricsDataChart.tsx:83]: You could get the `system` search param and use that to remove the system name.

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 3 above)
Comment (ivan-kishko) [reply publisher/src/components/DataViz/MetricsDataChart.tsx:384]: yeah it is one, we have to put ref on chart component but the button and function which are using this ref is inside data page

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 1 above)
Comment (ivan-kishko) [reply publisher/src/components/DataViz/MetricsDataChart.tsx]: done

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 2 above)
Comment (mxosman) [reply publisher/src/components/DataViz/MetricsDataChart.tsx]: Any luck or are you still seeing those errors?

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 1 above)
Comment (mxosman) [reply publisher/src/components/DataViz/MetricsDataChart.tsx]: Amazing - thank you so much!

**Review by mxosman — APPROVED** [review]
LGTM - thank you, Ivan!

File: publisher/src/components/DataViz/MetricsDataChart.tsx (hunk 2 above)
Comment (ivan-kishko) [reply publisher/src/components/DataViz/MetricsDataChart.tsx]: for some reason I do not even though I reloaded my webstorm before, today for some reason it accepts those imports)

Comment (mxosman) [reply publisher/src/components/DataViz/MetricsDataChart.tsx]: Yeah - it happens to me too every once in a while. Must be some weird state in the IDE or something. Glad it's all good now.

**Pull request merged**